### PR TITLE
Unique program names

### PR DESCRIPTION
**Pull request by borna223**
Program names have to be unique from now on.

### CODE REVIEW

File: server/program/program.model.js [hunk 1]
```diff
@@ -1,15 +1,18 @@
 'use strict';
 
-const { Model } = require('sequelize');
+const { Model, Sequelize } = require('sequelize');
+const trim = require('lodash/trim');
 
+const { Op } = Sequelize;
```
Comment (vladimyr) [inline server/program/program.model.js]: You can destructure `Op` from `require('sequelize');` :bowtie:

File: server/program/program.model.js [hunk 2]
```diff
@@ -1,15 +1,18 @@
...
 class Program extends Model {
   static fields(DataTypes) {
     return {
       name: {
         type: DataTypes.STRING,
         allowNull: false,
         unique: true,
-        validate: { notEmpty: true, len: [2, 255] }
+        validate: { notEmpty: true, len: [2, 255] },
+        set(value) { this.setDataValue('name', trim(value)); }
```
Comment (vladimyr) [inline server/program/program.model.js]: Ok, but why? You could set default value to make sure it is string and just use native trim? 🤔

**Review by vladimyr — CHANGES_REQUESTED** [review]

File: client/common/validation.js [hunk 3]
```diff
@@ -9,7 +10,20 @@ const alphanumerical = {
   }
 };
 
+const uniqueProgramName = {
+  getMessage: (field, args, data) => `Program named "${data}" already exists.`,
+  validate: async (name, [program]) => {
+    const [fetchedProgram] = await api.fetch({ params: { name, deleted: true } });
```
Comment (vladimyr) [inline client/common/validation.js:17]: You don't need to fetch if program name is same as initial value:
```js
if (program && program.name === name) return true;
```
which consequentially makes id checking obsolete 😉

File: server/program/program.controller.js [hunk 4]
```diff
@@ -3,10 +3,15 @@
...
 
+const { Op } = Sequelize;
+
 const processInput = input => pick(input, ['name', 'startDate', 'endDate']);
 
-function list(req, res) {
-  return Program.findAll().then(programs => res.jsend.success(programs));
+function list({ query: { name, deleted } }, res) {
+  const where = {};
+  if (name) where.name = { [Op.iLike]: name };
```
Comment (vladimyr) [inline server/program/program.controller.js]: This should be trimmed? 🤔

**Review by vladimyr — CHANGES_REQUESTED** [review]

File: client/common/validation.js (hunk 3 above)
Comment (borna223) [reply client/common/validation.js:17]: But what if new program name differs only in case sensitivity?

Comment (vladimyr) [reply client/common/validation.js:17]: In that case do case insensitive comparison?

**Pull request merged**